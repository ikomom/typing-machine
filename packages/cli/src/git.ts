import { relative, resolve } from 'path'
import { Snapshots } from 'ik-typing-machine'
import type { Snapshot } from 'ik-typing-machine'
import Git from 'simple-git'

/**
 * load form git history
 * @param path
 * @param cwd
 */
export async function loadFromGit(path: string, cwd = process.cwd()) {
  const git = Git(cwd)
  const root = (await git.raw(['rev-parse', '--show-toplevel'])).trim()
  const full = resolve(process.cwd(), path)
  const file = relative(root, full)
  console.log({ full, file, root })
  const items = await git.log({ file: full })
  // console.log({ items })

  const snaps = await Promise.all(items.all.map(async (arr): Promise<Snapshot> => {
    let content = ''
    try {
      content = await git.show(`${arr.hash}:${file}`)// :${file}
      console.log(arr.hash)
    }
    catch (e) {
      // console.log(e)
    }
    return {
      content,
    }
  }))
  // console.log({ snaps })

  return new Snapshots(...snaps)
}
