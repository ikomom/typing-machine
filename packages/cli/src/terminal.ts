import highlight from 'cli-highlight'
import type { Snapshots } from 'ik-typing-machine'
import cliCursor from 'cli-cursor'
import { createLogUpdate } from 'log-update'

const langMap: Record<string, string> = {
  ts: 'typescript',
}

export async function playInTerminal(snaps: Snapshots, lang?: string) {
  console.clear()
  const log = createLogUpdate(
    process.stdout,
    { showCursor: true },
  )

  for await (const snap of snaps.typeMachine()) {
    switch (snap.type) {
      case 'init':
      case 'insert':
      case 'removal':
      case 'animator-finish':
        cliCursor.hide()
        process.stdout.cursorTo(0, snap.content.split('\n').length - 1)
        log(highlight(snap.content, {
          language: lang ? langMap[lang] : lang,
          ignoreIllegals: true,
        }))

        if (snap.type !== 'init' && snap.type !== 'animator-finish') {
          const pre = snap.content.slice(0, snap.cursor)
          const lines = pre.split('\n')
          const char = lines[lines.length - 1].length
          process.stdout.cursorTo(char, lines.length - 1)
          cliCursor.show()
        }
    }
  }
  log.done()
}
