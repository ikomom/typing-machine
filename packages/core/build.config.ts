import { defineBuildConfig } from 'unbuild'

export default defineBuildConfig({
  entries: [
    'src/index',
  ],
  declaration: true,
  clean: true,
  externals: [
    // 'diff-match-patch-typescript',
  ],
  rollup: {
    emitCJS: true,
    // inlineDependencies: true,
  },
})
