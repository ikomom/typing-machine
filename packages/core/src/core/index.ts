export * from './patch'
export * from './animator'
export * from './types'
export * from './snaps'
