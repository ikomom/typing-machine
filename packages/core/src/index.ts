export * from './patch'
export * from './animator'
export * from './types'
export * from './snaps'
export * from './utils'
export * from './slicing'
export * from './timming'
