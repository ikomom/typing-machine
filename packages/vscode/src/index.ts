import { existsSync, promises as fs } from 'fs'
import { Range, Selection, commands, window, workspace } from 'vscode'
import { SnapshotManager, Snapshots, simpleAnimator } from 'ik-typing-machine'
import { logOut } from './log'

const snapExt = '.typingmachine'
// const snapExt = '.ts'

function sleep(ms: number) {
  return new Promise(resolve => setTimeout(resolve, ms))
}
function getSnapPath(id: string) {
  return id + snapExt
}

async function writeSnapshots(path: string, snap: Snapshots) {
  const filepath = getSnapPath(path)
  await fs.writeFile(filepath, snap.toString(), 'utf-8')
}

export function activate() {
  logOut('active plugin')

  const manager = new SnapshotManager({
    // if old snapshot file exist, read it
    async ensureFallback(path: string) {
      const filepath = getSnapPath(path)
      if (existsSync(filepath)) {
        const content = await fs.readFile(filepath, 'utf-8')
        const snaps = Snapshots.fromString(content)
        window.showInformationMessage('Snapshots loaded from file')
        return snaps
      }
    },
  })

  const wacther = workspace.createFileSystemWatcher(`**/*${snapExt}`)
  // const wacther = workspace.createFileSystemWatcher('**/*.ts')
  // delete old snaps when recreate or rename file
  wacther.onDidCreate((uri) => {
    logOut('File:onDidCreate', uri)
    manager.delete(uri.path.replace(snapExt, ''))
  })
  wacther.onDidChange((uri) => {
    logOut('File:onDidChange', uri)
  })

  commands.registerCommand('ik-typing-machine.snap', async () => {
    const doc = window.activeTextEditor?.document
    if (!doc) {
      window.showWarningMessage('document not open')
      return
    }
    const path = doc.uri.fsPath
    const snaps = await manager.ensure(path)
    snaps.push({
      content: doc.getText(),
    })
    logOut('snap', { path, snaps })

    await writeSnapshots(path, snaps)

    window.showInformationMessage(`succeed take ${snaps.length} snapShot of ${doc.fileName}`)
  })

  commands.registerCommand('ik-typing-machine.play', async () => {
    const editor = window.activeTextEditor
    const doc = editor?.document
    if (!doc || !editor) {
      window.showWarningMessage('document not open')
      return
    }
    const snaps = await manager.ensure(doc.uri.fsPath)
    if (!snaps.length) {
      window.showWarningMessage('No snaps found')
      return
    }
    // compare lastSnap with current doc
    const lastSnap = snaps[snaps.length - 1]
    if (lastSnap.content !== doc.getText()) {
      const take = 'Take Snapshot'
      const discard = 'Discard'
      const cancel = 'Cancel'

      const result = await window.showInformationMessage(
        'The current document has been modified since last snapshot. Do you want take another snapshot?',
        { modal: true },
        take,
        discard,
        cancel,
      )
      if (!result || result === cancel)
        return

      if (result === take)
        await commands.executeCommand('ik-typing-machine.snap')
    }
    window.showInformationMessage(`Playing ${doc.fileName}`)

    let lastContent: string | undefined
    for (const snap of snaps) {
      // save first exist snap document
      if (lastContent === undefined) {
        lastContent = snap.content
        // clear doc
        await editor.edit((edit) => {
          edit.replace(new Range(0, 0, Infinity, Infinity), lastContent!)
        })
        continue
      }

      // play animate
      const animator = simpleAnimator(lastContent!, snap.content)
      for (const result of animator) {
        const pos = doc.positionAt(result.cursor)
        await editor.edit((edit) => {
          if (result.type === 'insert-ing')
            edit.insert(doc.positionAt(result.cursor - 1), result.char ?? '')
          else if (result.type === 'removal-ing')
            edit.delete(new Range(pos, doc.positionAt(result.cursor + 1)))
        })

        editor.selection = new Selection(pos, pos)

        await sleep(Math.random() * 200)
      }

      lastContent = snap.content
    }

    window.showInformationMessage(`Finished ${doc.fileName}`)
  })
}

export function deactivate() {
  logOut('deactivate plugin')
}
