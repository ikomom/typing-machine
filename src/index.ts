import type { Diff } from 'diff-match-patch-typescript'
import { DiffMatchPatch, DiffOperation } from 'diff-match-patch-typescript'

export interface InsertPatch {
  type: 'insert'
  // 删除开始位置
  from: number
  text: string
}
export interface RemovalPatch {
  type: 'removal'
  // 删除开始位置
  from: number
  length: number
}

export type Patch = InsertPatch | RemovalPatch

export function calculatePatches(diffs: Diff[]): Patch[] {
  const patches: Patch[] = []
  // 当前文字的index
  let cursor = 0
  for (const change of diffs) {
    const [operation, text] = change
    const offset = text.length
    if (operation === DiffOperation.DIFF_EQUAL) {
      cursor += offset
    }
    else if (operation === DiffOperation.DIFF_DELETE) {
      patches.push({
        type: 'removal',
        from: cursor,
        length: offset,
        // // eslint-disable-next-line @typescript-eslint/ban-ts-comment
        // // @ts-expect-error
        // text, // TODO: 测试用, 待移除
      })
      // 删除时，游标还是一样，所以不用移动，一下代码不用
      // cursor -= offset
    }
    else if (operation === DiffOperation.DIFF_INSERT) {
      patches.push({
        type: 'insert',
        from: cursor,
        text,
      })
      // 添加时, 游标 = 之前游标 + 添加长度
      cursor += offset
    }
    else {
      throw new Error('unknown change type')
    }
  }
  return patches
}

export function diffString(a: string, b: string): Diff[] {
  const diff = new DiffMatchPatch()
  const res = diff.diff_main(a, b)
  diff.diff_cleanupSemantic(res)
  return res
}

export function applyPatches(input: string, patches: Patch[]) {
  let output = input
  // 每次只关注添加、删除的位置, 并且拼接起来
  for (const patch of patches) {
    const head = output.slice(0, patch.from)
    if (patch.type === 'insert') {
      const tail = output.slice(patch.from)
      output = head + patch.text + tail
    }
    else if (patch.type === 'removal') {
      const tail = output.slice(patch.from + patch.length)
      output = head + tail
    }
  }
  return output
}
