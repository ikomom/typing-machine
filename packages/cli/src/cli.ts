import { parse as parsePath, resolve } from 'path'
import { promises as fs } from 'fs'
import { Snapshots, getSnapshotPath } from 'ik-typing-machine'
import cac from 'cac'
import { version } from '../package.json'
import { loadFromGit } from './git'
import { playInTerminal } from './terminal'

const cli = cac('typemachine')

cli.version(version)
  .option('-r, --root <path>', 'root path')
  .option('--git', 'Read from Git history')
  .help()

cli.command('play <file>').action(play)

cli
  .command('')
  .action(() => {
    cli.outputHelp()
  })
cli.parse()

export interface CliOptions {
  git?: string
  root?: string
}

async function play(file: string, options: CliOptions) {
  const { root = process.cwd(), git } = options
  const absolute = resolve(root, file)

  const snaps = git ? await loadFromGit(absolute, root) : await loadFromFile(absolute)
  const { ext } = parsePath(file)

  console.log({ file, options, absolute, snaps: snaps.length, ext })
  await playInTerminal(snaps, ext.slice(1))
}

async function loadFromFile(file: string) {
  return Snapshots.fromString(await fs.readFile(getSnapshotPath(file), 'utf-8'))
}
