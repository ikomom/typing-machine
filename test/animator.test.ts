import { describe, expect, it } from 'vitest'
import { calculatePatches, createAnimator, diffString } from '../packages/core/src'
import { input, output } from './fixture'

describe('animator', () => {
  it('generate', () => {
    const delta = diffString(input, output)
    const patches = calculatePatches(delta)
    const animator = createAnimator(input, patches)
    expect([...animator]).toMatchSnapshot('animator')
  })
})
