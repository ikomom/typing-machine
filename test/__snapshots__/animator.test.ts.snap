// Vitest Snapshot v1

exports[`animator > generate > animator 1`] = `
[
  {
    "cursor": 71,
    "output": "
import { describe, expect, it } from 'vitest'
import {ll} from './one'

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hello')
  })
})
",
    "type": "removal-start",
  },
  {
    "cursor": 71,
    "output": "
import { describe, expect, it } from 'vitest'
import {ll} from './one'

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hello')
  })
})
",
    "type": "removal-ing",
  },
  {
    "cursor": 70,
    "output": "
import { describe, expect, it } from 'vitest'
import {ll} from './one

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hello')
  })
})
",
    "type": "removal-ing",
  },
  {
    "cursor": 69,
    "output": "
import { describe, expect, it } from 'vitest'
import {ll} from './on

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hello')
  })
})
",
    "type": "removal-ing",
  },
  {
    "cursor": 68,
    "output": "
import { describe, expect, it } from 'vitest'
import {ll} from './o

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hello')
  })
})
",
    "type": "removal-ing",
  },
  {
    "cursor": 67,
    "output": "
import { describe, expect, it } from 'vitest'
import {ll} from './

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hello')
  })
})
",
    "type": "removal-ing",
  },
  {
    "cursor": 66,
    "output": "
import { describe, expect, it } from 'vitest'
import {ll} from '.

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hello')
  })
})
",
    "type": "removal-ing",
  },
  {
    "cursor": 65,
    "output": "
import { describe, expect, it } from 'vitest'
import {ll} from '

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hello')
  })
})
",
    "type": "removal-ing",
  },
  {
    "cursor": 64,
    "output": "
import { describe, expect, it } from 'vitest'
import {ll} from 

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hello')
  })
})
",
    "type": "removal-ing",
  },
  {
    "cursor": 63,
    "output": "
import { describe, expect, it } from 'vitest'
import {ll} from

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hello')
  })
})
",
    "type": "removal-ing",
  },
  {
    "cursor": 62,
    "output": "
import { describe, expect, it } from 'vitest'
import {ll} fro

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hello')
  })
})
",
    "type": "removal-ing",
  },
  {
    "cursor": 61,
    "output": "
import { describe, expect, it } from 'vitest'
import {ll} fr

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hello')
  })
})
",
    "type": "removal-ing",
  },
  {
    "cursor": 60,
    "output": "
import { describe, expect, it } from 'vitest'
import {ll} f

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hello')
  })
})
",
    "type": "removal-ing",
  },
  {
    "cursor": 59,
    "output": "
import { describe, expect, it } from 'vitest'
import {ll} 

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hello')
  })
})
",
    "type": "removal-ing",
  },
  {
    "cursor": 58,
    "output": "
import { describe, expect, it } from 'vitest'
import {ll}

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hello')
  })
})
",
    "type": "removal-ing",
  },
  {
    "cursor": 57,
    "output": "
import { describe, expect, it } from 'vitest'
import {ll

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hello')
  })
})
",
    "type": "removal-ing",
  },
  {
    "cursor": 56,
    "output": "
import { describe, expect, it } from 'vitest'
import {l

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hello')
  })
})
",
    "type": "removal-ing",
  },
  {
    "cursor": 55,
    "output": "
import { describe, expect, it } from 'vitest'
import {

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hello')
  })
})
",
    "type": "removal-ing",
  },
  {
    "cursor": 54,
    "output": "
import { describe, expect, it } from 'vitest'
import 

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hello')
  })
})
",
    "type": "removal-ing",
  },
  {
    "cursor": 53,
    "output": "
import { describe, expect, it } from 'vitest'
import

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hello')
  })
})
",
    "type": "removal-ing",
  },
  {
    "cursor": 52,
    "output": "
import { describe, expect, it } from 'vitest'
impor

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hello')
  })
})
",
    "type": "removal-ing",
  },
  {
    "cursor": 51,
    "output": "
import { describe, expect, it } from 'vitest'
impo

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hello')
  })
})
",
    "type": "removal-ing",
  },
  {
    "cursor": 50,
    "output": "
import { describe, expect, it } from 'vitest'
imp

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hello')
  })
})
",
    "type": "removal-ing",
  },
  {
    "cursor": 49,
    "output": "
import { describe, expect, it } from 'vitest'
im

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hello')
  })
})
",
    "type": "removal-ing",
  },
  {
    "cursor": 48,
    "output": "
import { describe, expect, it } from 'vitest'
i

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hello')
  })
})
",
    "type": "removal-ing",
  },
  {
    "cursor": 47,
    "output": "
import { describe, expect, it } from 'vitest'


describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hello')
  })
})
",
    "type": "removal-ing",
  },
  {
    "cursor": 46,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hello')
  })
})
",
    "type": "removal-ing",
  },
  {
    "cursor": 71,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hello')
  })
})
",
    "type": "removal-end",
  },
  {
    "cursor": 138,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hello')
  })
})
",
    "type": "removal-start",
  },
  {
    "cursor": 138,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hello')
  })
})
",
    "type": "removal-ing",
  },
  {
    "cursor": 137,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hello)
  })
})
",
    "type": "removal-ing",
  },
  {
    "cursor": 136,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hell)
  })
})
",
    "type": "removal-ing",
  },
  {
    "cursor": 135,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('hel)
  })
})
",
    "type": "removal-ing",
  },
  {
    "cursor": 134,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('he)
  })
})
",
    "type": "removal-ing",
  },
  {
    "cursor": 133,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log('h)
  })
})
",
    "type": "removal-ing",
  },
  {
    "cursor": 132,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log(')
  })
})
",
    "type": "removal-ing",
  },
  {
    "cursor": 131,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log()
  })
})
",
    "type": "removal-ing",
  },
  {
    "cursor": 130,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.log)
  })
})
",
    "type": "removal-ing",
  },
  {
    "cursor": 129,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.lo)
  })
})
",
    "type": "removal-ing",
  },
  {
    "cursor": 128,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.l)
  })
})
",
    "type": "removal-ing",
  },
  {
    "cursor": 127,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33)
    console.)
  })
})
",
    "type": "removal-ing",
  },
  {
    "cursor": 126,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33)
    console)
  })
})
",
    "type": "removal-ing",
  },
  {
    "cursor": 125,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33)
    consol)
  })
})
",
    "type": "removal-ing",
  },
  {
    "cursor": 124,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33)
    conso)
  })
})
",
    "type": "removal-ing",
  },
  {
    "cursor": 123,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33)
    cons)
  })
})
",
    "type": "removal-ing",
  },
  {
    "cursor": 122,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33)
    con)
  })
})
",
    "type": "removal-ing",
  },
  {
    "cursor": 121,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33)
    co)
  })
})
",
    "type": "removal-ing",
  },
  {
    "cursor": 120,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33)
    c)
  })
})
",
    "type": "removal-ing",
  },
  {
    "cursor": 119,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33)
    )
  })
})
",
    "type": "removal-ing",
  },
  {
    "cursor": 118,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33)
   )
  })
})
",
    "type": "removal-ing",
  },
  {
    "cursor": 117,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33)
  )
  })
})
",
    "type": "removal-ing",
  },
  {
    "cursor": 116,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33)
 )
  })
})
",
    "type": "removal-ing",
  },
  {
    "cursor": 115,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33)
)
  })
})
",
    "type": "removal-ing",
  },
  {
    "cursor": 114,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33))
  })
})
",
    "type": "removal-ing",
  },
  {
    "cursor": 138,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33))
  })
})
",
    "type": "removal-end",
  },
  {
    "cursor": 114,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33))
  })
})
",
    "type": "insert-start",
  },
  {
    "cursor": 115,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33).)
  })
})
",
    "type": "insert-ing",
  },
  {
    "cursor": 116,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33).t)
  })
})
",
    "type": "insert-ing",
  },
  {
    "cursor": 117,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33).to)
  })
})
",
    "type": "insert-ing",
  },
  {
    "cursor": 118,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33).toE)
  })
})
",
    "type": "insert-ing",
  },
  {
    "cursor": 119,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33).toEq)
  })
})
",
    "type": "insert-ing",
  },
  {
    "cursor": 120,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33).toEqu)
  })
})
",
    "type": "insert-ing",
  },
  {
    "cursor": 121,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33).toEqua)
  })
})
",
    "type": "insert-ing",
  },
  {
    "cursor": 122,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33).toEqual)
  })
})
",
    "type": "insert-ing",
  },
  {
    "cursor": 123,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33).toEqual()
  })
})
",
    "type": "insert-ing",
  },
  {
    "cursor": 124,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33).toEqual(1)
  })
})
",
    "type": "insert-ing",
  },
  {
    "cursor": 114,
    "output": "
import { describe, expect, it } from 'vitest'

describe('should', () => {
  it('exported', () => {
    expect(33).toEqual(1)
  })
})
",
    "type": "insert-end",
  },
]
`;
